[subin] Vote Contract, NFT Contract 작성

anxiubin opened the pull request.

anxiubin committed:
✨ KIP17 표준 컨트랙트 추가

anxiubin committed:
✨ KIP17MetadataMintable 컨트랙트 수정

- mint 할 때 NFTHolders 에 매핑하는 함수 추가
- NFT 홀더인지 , 마스터 NFT 홀더인지 체크하는 함수 추가

anxiubin committed:
✨ Klaytn17MintBadgemeal 컨트랙트 추가

anxiubin committed:
♻️ Vote contract 개선_v.1.0

- Ownable 컨트랙트 상속: onlyOwner사용할 수 있도록 추가
- proposeMenu 함수에서 KIP17MetadataMintable 컨트랙트의 isMasterNFTHolder함수를 이용해서 메뉴 추가 권한 검증
- vote 함수에서  KIP17MetadataMintable 컨트랙트의 isNFTHolder함수를 이용해서 투표 권한 검증
- addWinnerProposal 함수 추가 : 투표가 마감되는 시점에 호출해야함.

anxiubin committed:
♻️ Vote contract 개선_v.1.1

1. KIP17 contract 개선
- getOwnedTokens 추가
- _tokenLevel 매핑 생성
- tokenLevel 추가: tokenId에 매핑되는 level 리턴
- mintWithTokenURI 함수에 _setTokenLevel(tokenId, level) 추가해서 토큰마다 level 매핑

2. Vote Contract 개선
- isNFTholder 함수 추가 : getOwnedTokens 함수 이용
- isMasterNFTholder 함수 추가: tokenLevel 함수 이용

anxiubin committed:
🐛 result 리턴

anxiubin committed:
♻️ vote 함수에 require 문 추가 및 sender 변수 제거

anxiubin committed:
🐛 메뉴 추가에 imageUrl은 필요없으므로 삭제

anxiubin committed:
💬 README 업데이트

anxiubin committed:
♻️ 토큰을 삭제할 때 _tokenLevel도 함께 삭제

anxiubin committed:
💬 README 업데이트

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+## KIP17Metadata 컨트랙트 수정 사항
+
+1. _tokenLevel 매핑 추가
+일반 NFT와 마스터 NFT를 구분하기 위해 `_tokenLevel` 매핑을 추가해서 활용했다.
```
NFT 종류를 구분하는 변수가 필요했는데 잘 넣으신 것 같습니다!👍

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+
+    // Clear metadata (if any)
+    if (bytes(_tokenURIs[tokenId]).length != 0) {
+        delete _tokenURIs[tokenId];
```
bytes() 함수로 매핑 값이 존재하는지를 체크하는 방법이 있었네요!
잘 배우고 갑니다.👍

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+    // 🔥 Clear level 
+    if (bytes(_tokenLevel[tokenId]).length != 0) {
+        delete _tokenLevel[tokenId];
+    }
```
bytes() 함수에 uint값이 들어가도 컴파일 때 문제가 안생겼나요??
궁금해서 ide에서 테스트해봤는데
TypeError: Explicit type conversion not allowed from "uint256" to "bytes memory".
        return bytes(_tokenLevel[tokenId]).length;
 라고 뜨네요..ㅠ

anxiubin replied:
앗 제가 수정을 안하고 올렸네요! 아래와 같이 코드 바꾸시면 컴파일 됩니다!
```sol
        if (_tokenLevel[tokenId]) {
            delete _tokenLevel[tokenId];
        }
```

anxiubin replied:
앗 제가 수정을 안하고 올렸네요! 아래와 같이 코드 바꾸시면 컴파일 됩니다!
토큰레벨은 1 또는 2 이므로 아래와 같이 조건문을 걸면 될 것 같아요.
```sol
       if (_tokenLevel[tokenId] > 0) {
            delete _tokenLevel[tokenId];
        }```

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+    _mint(to, tokenId);
+    _setTokenURI(tokenId, tokenURI);
+    _setTokenLevel(tokenId, level);
+}
```
깔끔한 코드입니다.👍

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+1. Ownable 컨트랙트를 가져온다. 
+@openzeppelin/contracts/ownership/Ownable.sol
+
+2. 뱃지밀 컨트랙트에 Ownable 컨트랙트를 상속 받는다.
```
Ownable 컨트랙트를 상속해서 소유권 확인을 하는 방법이 있었네요!
잘 배우고 갑니다.👍

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+struct Proposal {
+  string name;   // 메뉴 이름
+  uint voteCount; // 투표 받은 수
+  string imageUrl; // 메뉴 이미지 url
```
https://github.com/BadgeMeal/badgemeal-contract/pull/4#discussion_r805792849
리뷰 내용과 같이 없애는 것이 맞는 것 같습니다!

anxiubin replied:
컨트랙트 코드에서는 삭제했는데 README에서 안지웠네요😂

sapcy commented on README.md:
````diff
@@ -1,3 +1,196 @@
+```sol
+function vote(uint _proposal, address _nftAddress) public {
+    require(isNFTholder(_nftAddress), "You have no right to vote");
+    require(!voters[msg.sender].voted, "Already voted.");
````
voters[msg.sender] 값이 null인지 확인하는 require문(투표자 존재 여부)도 필요할 것 같습니다.

anxiubin replied:
저는 voters 에 값을 미리 매핑하지않고, vote 함수를 호출할 때 매핑하고 있습니다! 그래서 첫 투표자는 아래 조건을 통과하고 vote함수를 실행하며 매핑 됩니다.
```sol
require(isNFTholder(_nftAddress), "You have no right to vote");
require(!voters[msg.sender].voted, "Already voted.");
```

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+- 기존 메뉴 리스트는 DB에 있고, 투표로 추가된 메뉴 리스트는 vote 컨트랙트의 `winnerProposals`변수에 담긴다.
+- 호츨 조건: 투표가 마감되는 시점에 백엔드에서 호출한다.
+- 백엔드에서 DB의 메뉴 리스트에 `winnerProposals`를 추가한다.
+- ‼️보완해야할 사항 : voters를 초기화하는 방법을 찾아봐야한다.
```
저도 이 부분이 애매해서 투표마다 컨트랙트를 새로 배포하는 걸 생각했습니다.
지금 지식으로는 voters를 loop해서 delete하는 방법밖에 없어서..
회의 때 얘기해보면 좋을 것 같아요!

anxiubin replied:
voteStartTime과 투표자 매핑하는 방식을 써서 반복문을 돌지 않아도 되도록 수정했습니다 :)

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+  uint vote;   // Menu 리스트 요소의 index (0,1,2 ...)
+}
+  
+mapping(address => Voter) public voters; // 투표자 매핑
```
voters 매핑에 투표자를 추가하는 함수가 없는 것 같습니다.

anxiubin replied:
vote 함수를 실행하며 매핑합니다!

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+    uint256[] memory ownedTokenLIst = Klaytn17MintBadgemeal(_nftAddress).getOwnedTokens(msg.sender);
+    bool result = false;
+    for (uint256 i = 0; i < ownedTokenLIst.length; i++) {
+        if (Klaytn17MintBadgemeal(_nftAddress).tokenLevel(ownedTokenLIst[i]) == 2) {
```
이 로직대로면 ownedTokenLIst가 많으면 많을 수록 가스비가 증가할 것이므로 보완이 필요해 보입니다.
방기진 강사님이 추천하신 방식으로 바꾸는 것도 좋은 방법일 것 같아요!
- 강사님 예) mapping(address => boolean) isMasterTokenMap;

anxiubin replied:
view 속성이 있는 함수는 가스비가 들지 않는다고 알고 있습니다! 따라서 해당 함수는 가스비 문제는 없을 것으로 판단됩니다.

anxiubin committed:
💬 README 업데이트

sapcy commented on README.md:
```diff
@@ -1,3 +1,196 @@
+    name: winner.name,
+    proposer: winner.proposer,
+    voteCount: winner.voteCount
+    }));
```
Proposal storage winner를 선언한 부분에서 이미 winner에 메모리 할당이 되었으므로  winnerProposals.push(winner) 처럼 storage 선언된 메모리 포인터를 그대로 넣어줘야 될 것 같습니다. 
위의 코드대로면 구조체 생성자 Proposal({}) 로 새로 할당된 포인터(위 winner변수 포인터와 다름)가 push되는 것으로 보이네요. 
(솔리디티 자료구조 지식이 부족하여 틀렸을 수 있습니다.)

anxiubin replied:
아 그부분을 놓쳤네요! 시영님이 말씀하신대로 아래와 같이 반영했습니다 👍
```sol
Proposal storage winner = proposals[winningProposal()];
winnerProposals.push(winner);
```

sapcy reviewed (commented):
컨트랙트 작성하느라 고생하셨습니다!👍👍

Hyunki6040 commented on README.md:
```diff
@@ -1,3 +1,200 @@
+
+7. 가장 많은 득표수를 얻은 메뉴를 향후 메뉴 리스트에 추가하는 함수
+- 기존 메뉴 리스트는 DB에 있고, 투표로 추가된 메뉴 리스트는 vote 컨트랙트의 `winnerProposals`변수에 담긴다.
+- 호츨 조건: 투표가 마감되는 시점에 백엔드에서 호출한다.
```
투표가 마감되는 시점을 owner가 해당하는 function을 호출하지 않더라도 종료일자가 지났을 때 마감되도록 하면 이상적일 것 같아요! 아무래도 블록체인에 관리자가 개입하는 경우의 수는 최대한 줄이는게 좋다고 생각해요. 제 코드부분에서는 modifier로 선언해서 결과를 조회할 때마다 trigger로 실행되도록 했는데 코드가 그렇게 효율적인 느낌이 들지는 않더라고요,,! 한번 고민해보면 좋을 것 같아요.

anxiubin replied:
투표 시간을 설정하고 투표 마감 시간은 자동으로 세팅되도록 코드 수정 했습니다!
```sol
uint public proposeStartTime; // 메뉴 추가 시작 시간
uint public voteStartTime; // 투표 시작 시간

// 메뉴 추가 가능한 시간인지 검증
modifier proposeAvailable() {
    require(now >= proposeStartTime && now < proposeStartTime + 1 days, "Cannot propose now.");
    _;
}
// 투표 가능한 시간인지 검증
modifier voteAvailable() {
    require(now >= voteStartTime && now < voteStartTime + 1 days, "Cannot vote now.");
    _;
}

```

Hyunki6040 commented on README.md:
```diff
@@ -1,3 +1,200 @@
+// - 호출 조건: 투표가 마감되는 시점.
+function addWinnerProposal() public onlyOwner {
+    Proposal storage winner = proposals[winningProposal()];
+
```
투표가 마감되는 시점에 require로 투표 종료시간 이후를 확인해주면 더 좋을 것 같아요

anxiubin replied:
채택된 메뉴를 추가 가능한지 검증하는 modifier를 추가했어요! 현재 시간이 투표 마감 시간 이후면 가능하도록 수정했습니다.
```sol
// 채택된 메뉴를 추가 가능한 시간인지 검증
modifier addProposeAvailable() {
  require(now > voteStartTime + 1 days, "Voting is currently underway.");
    _;
}

function addWinnerProposal(address _nftAddress) public onlyOwner addProposeAvailable {
}

```

Hyunki6040 reviewed (commented):
주석도 이해하기 쉽게 잘달아 주시고 고생하셨어요!

Yunseongpyo commented:
고생하셨습니다👍

anxiubin committed:
♻️ NFT 컨트랙트 , Vote 컨트랙트 수정

- tokenLevel 함수 속성 수정
- burn 할때 _tokenLevel 조건문 수정
- voters 초기화를 위해 votersAddressList 추가
- addWinnerProposal 함수 실행시 AddWinner event 발생
- addWinnerProposal 함수 수정

anxiubin committed:
♻️ addWinnerProposal require문 수정: NFT 홀더 과반수 이상

anxiubin committed:
🐛 세미콜론 누락된 것 추가

anxiubin committed:
✨vote 컨트랙트에서 투표 수 증가시킬 때 safemath 라이브러리 사용

anxiubin committed:
♻️ Vote 컨트랙트 보완 : voters 초기화 반복문 삭제, 시간 함수 추가

- voteStartTime과 투표자 매핑하여 voters 초기화를 위해 반복문을 돌지 않도록 수정
- 투표 시작, 메뉴 추가 시작 시간을 세팅하는 함수, 시간 검증하는 함수변경자 추가
- safemath 라이브러리 관련 오류 수정

anxiubin committed:
♻️ 투표 마감 시간 이후에 채택된 메뉴를 추가할 수 있도록 수정

anxiubin committed:
💬 README 업데이트

anxiubin committed:
♻️ NFT 컨트랙트 보완 및 Vote 컨트랙트 수정

- _nftType, _menuType 추가
- _burnForMasterNFT 함수 추가
- KIP17MetadataMintable 컨트랙트에 KIP17Enumerable 상속
- mintWithTokenURI 수정 : 특정 NFT(ex: 국밥 NFT)를 19개 이상 소유했는지 판별해서 19개를 삭제한 후 마스터 NFT 배지 발행하고, 아니라면 일반 메뉴 NFT 발행
- _checkMenu 함수 추가
- _removeOwnToken 함수 추가
- mintWithKlay 함수 추가

anxiubin committed:
♻️ mintWithKlay 함수에서 receiver 세팅하는 방식 보완

anxiubin committed:
♻️ genralTokenURI와 masterTokenURI를 구분해서 인자로 넘겨받도록 수정

anxiubin committed:
♻️ MintMasterNFT event 추가 및 주석 수정

anxiubin committed:
✨ mintWithTokenURI를 호출할 때 필요한 onlyBadgemealMinter 함수변경자 추가

- onlyMinter 대신 함수변경자를 뱃지밀 프로젝트에 맞춰서 다시 세팅

anxiubin committed:
♻️ onlyBadgemealMinter에 address 넣어서 체크하도록 수정

anxiubin committed:
✨ proposals length 리턴하는 함수 추가

anxiubin committed:
✨ winnerProposals 길이 가져오는 함수 추가

anxiubin committed:
💬 리드미 업데이트
